File: src/main/java/core/db/FaktorenTable.java
package core.db;

import core.model.FactorObject;
import core.model.FormulaFactors;
import java.sql.Types;


public final class FaktorenTable extends AbstractTable {

	/** tablename **/
	public final static String TABLENAME = "FAKTOREN";
	
	FaktorenTable(ConnectionManager adapter){
		super(TABLENAME, adapter);
	}

	@Override
	protected void initColumns() {
		columns = new ColumnDescriptor[]{
				ColumnDescriptor.Builder.newInstance().setColumnName("PositionID").setGetter((o) -> ((FactorObject) o).getPosition()).setSetter((o, v) -> ((FactorObject) o).setPosition((byte)(int) v)).setType(Types.INTEGER).isNullable(false).isPrimaryKey(true).build(),
				ColumnDescriptor.Builder.newInstance().setColumnName("GKfactor").setGetter((o) -> ((FactorObject) o).getGKfactor()).setSetter((o, v) -> ((FactorObject) o).setTorwart((float) v)).setType(Types.REAL).isNullable(false).build(),
				ColumnDescriptor.Builder.newInstance().setColumnName("DEfactor").setGetter((o) -> ((FactorObject) o).getDEfactor()).setSetter((o, v) -> ((FactorObject) o).setDefendingFactor((float) v)).setType(Types.REAL).isNullable(false).build(),
				ColumnDescriptor.Builder.newInstance().setColumnName("WIfactor").setGetter((o) -> ((FactorObject) o).getWIfactor()).setSetter((o, v) -> ((FactorObject) o).setWingerFactor((float) v)).setType(Types.REAL).isNullable(false).build(),
				ColumnDescriptor.Builder.newInstance().setColumnName("PSfactor").setGetter((o) -> ((FactorObject) o).getPSfactor()).setSetter((o, v) -> ((FactorObject) o).setPassingFactor((float) v)).setType(Types.REAL).isNullable(false).build(),
				ColumnDescriptor.Builder.newInstance().setColumnName("SPfactor").setGetter((o) -> ((FactorObject) o).getSPfactor()).setSetter((o, v) -> ((FactorObject) o).setSetPiecesFactor((float) v)).setType(Types.REAL).isNullable(false).build(),
				ColumnDescriptor.Builder.newInstance().setColumnName("SCfactor").setGetter((o) -> ((FactorObject) o).getSCfactor()).setSetter((o, v) -> ((FactorObject) o).setTorschuss((float) v)).setType(Types.REAL).isNullable(false).build(),
				ColumnDescriptor.Builder.newInstance().setColumnName("PMfactor").setGetter((o) -> ((FactorObject) o).getPMfactor()).setSetter((o, v) -> ((FactorObject) o).setPlaymakingFactor((float) v)).setType(Types.REAL).isNullable(false).build(),
				ColumnDescriptor.Builder.newInstance().setColumnName("NormalisationFactor").setGetter((o) -> ((FactorObject) o).getNormalizationFactor()).setSetter((o, v) -> ((FactorObject) o).setNormalizationFactor((float) v)).setType(Types.REAL).isNullable(false).build()
		};
	}

	@Override
	protected String createSelectStatement() {
		return createSelectStatement("");
	}

	void pushFactorsIntoDB(FactorObject fo) {
		if (fo != null) {
			store(fo);
		}
	}

	void getFaktorenFromDB() {
		var factors = load(FactorObject.class);
		if (!factors.isEmpty()) {
			for (var factor : factors) {
				FormulaFactors.instance().setPositionFactor(factor.getPosition(), factor);
			}
		} else {
			// use hardcoded values
			FormulaFactors.instance().importDefaults();
		}
	}
}

File: src/main/java/core/db/FaktorenTable.kt
package core.db

import core.model.FactorObject
import core.model.FormulaFactors
import java.sql.*
import java.util.function.BiConsumer
import java.util.function.Function

class FaktorenTable internal constructor(adapter: JDBCAdapter) : AbstractTable(TABLENAME, adapter) {
    override fun initColumns() {
        columns = arrayOf<ColumnDescriptor>(
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("PositionID")
                .setGetter(Function<Any?, Any?> { o: Any? -> (o as FactorObject?)!!.position }).setSetter(
                BiConsumer<Any?, Any> { o: Any?, v: Any -> (o as FactorObject?)!!.position = (v as Int).toByte() })
                .setType(
                    Types.INTEGER
                ).isNullable(false).isPrimaryKey(true).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("GKfactor")
                .setGetter(Function<Any?, Any?> { o: Any? -> (o as FactorObject?)!!.gKfactor }).setSetter(
                BiConsumer<Any?, Any> { o: Any?, v: Any -> (o as FactorObject?)!!.setTorwart(v as Float) })
                .setType(Types.REAL).isNullable(false).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("DEfactor")
                .setGetter(Function<Any?, Any?> { o: Any? -> (o as FactorObject?)!!.dEfactor }).setSetter(
                BiConsumer<Any?, Any> { o: Any?, v: Any -> (o as FactorObject?)!!.setDefendingFactor(v as Float) })
                .setType(
                    Types.REAL
                ).isNullable(false).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("WIfactor")
                .setGetter(Function<Any?, Any?> { o: Any? -> (o as FactorObject?)!!.wIfactor }).setSetter(
                BiConsumer<Any?, Any> { o: Any?, v: Any -> (o as FactorObject?)!!.setWingerFactor(v as Float) })
                .setType(
                    Types.REAL
                ).isNullable(false).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("PSfactor")
                .setGetter(Function<Any?, Any?> { o: Any? -> (o as FactorObject?)!!.pSfactor }).setSetter(
                BiConsumer<Any?, Any> { o: Any?, v: Any -> (o as FactorObject?)!!.setPassingFactor(v as Float) })
                .setType(
                    Types.REAL
                ).isNullable(false).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("SPfactor")
                .setGetter(Function<Any?, Any?> { o: Any? -> (o as FactorObject?)!!.sPfactor }).setSetter(
                BiConsumer<Any?, Any> { o: Any?, v: Any -> (o as FactorObject?)!!.setSetPiecesFactor(v as Float) })
                .setType(
                    Types.REAL
                ).isNullable(false).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("SCfactor")
                .setGetter(Function<Any?, Any?> { o: Any? -> (o as FactorObject?)!!.sCfactor }).setSetter(
                BiConsumer<Any?, Any> { o: Any?, v: Any -> (o as FactorObject?)!!.setTorschuss(v as Float) })
                .setType(Types.REAL).isNullable(false).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("PMfactor")
                .setGetter(Function<Any?, Any?> { o: Any? -> (o as FactorObject?)!!.pMfactor }).setSetter(
                BiConsumer<Any?, Any> { o: Any?, v: Any -> (o as FactorObject?)!!.setPlaymakingFactor(v as Float) })
                .setType(
                    Types.REAL
                ).isNullable(false).build(),
            ColumnDescriptor.Builder.Companion.newInstance().setColumnName("NormalisationFactor")
                .setGetter(Function<Any?, Any?> { o: Any? -> (o as FactorObject?)!!.normalizationFactor }).setSetter(
                BiConsumer<Any?, Any> { o: Any?, v: Any -> (o as FactorObject?)!!.normalizationFactor = v as Float })
                .setType(
                    Types.REAL
                ).isNullable(false).build()
        )
    }

    override fun createPreparedSelectStatementBuilder(): PreparedSelectStatementBuilder? {
        return PreparedSelectStatementBuilder(this, "")
    }

    fun pushFactorsIntoDB(fo: FactorObject?) {
        fo?.let { store(it) }
    }

    fun getFaktorenFromDB():Unit {
        val factors = load(FactorObject::class.java)
        if (factors.isNotEmpty()) {
            for (factor in factors) {
                FormulaFactors.instance().setPositionFactor(factor.position.toInt(), factor)
            }
        } else {
            // use hardcoded values
            FormulaFactors.instance().importDefaults()
        }
    }

    companion object {
        /** tablename  */
        const val TABLENAME = "FAKTOREN"
    }
}
